### PR TITLE
Add support for rhino as a compilation target

### LINKED ISSUE #2537: Cannot run standalone compiler from babel-core in Rhino JavaScript engine
```groovy
@Grab(group='org.mozilla', module='rhino', version='1.7.7')
import org.mozilla.javascript.Context
import org.mozilla.javascript.ContextFactory
import org.mozilla.javascript.ScriptableObject

ContextFactory contextFactory = new ContextFactory()
Context context = contextFactory.enterContext()
context.setOptimizationLevel(-1)
ScriptableObject scope = context.initStandardObjects()

new URL('https://cdnjs.cloudflare.com/ajax/libs/babel-core/5.8.25/browser.js').withReader {
  context.evaluateReader(scope, it, "browser.js", 1, null)
}
```
leads to
```
org.mozilla.javascript.EcmaError: TypeError: Expected argument of type object, but instead had type object (browser.js#56462)
	at org.mozilla.javascript.ScriptRuntime.constructError(ScriptRuntime.java:3951)
	at org.mozilla.javascript.ScriptRuntime.constructError(ScriptRuntime.java:3929)
	at org.mozilla.javascript.ScriptRuntime.typeError(ScriptRuntime.java:3962)
	at org.mozilla.javascript.ScriptRuntime.typeError1(ScriptRuntime.java:3974)
	at org.mozilla.javascript.ScriptableObject.ensureScriptable(ScriptableObject.java:2078)
	at org.mozilla.javascript.NativeObject.execIdCall(NativeObject.java:275)
	at org.mozilla.javascript.IdFunctionObject.call(IdFunctionObject.java:101)
	at org.mozilla.javascript.Interpreter.interpretLoop(Interpreter.java:1479)
	at script(browser.js:56462)
	at script.toType(browser.js:56399)
	at script(browser.js:56430)
	at script(browser.js:53410)
	at script.s(browser.js:2)
	at script(browser.js:2)
	at script(browser.js:57061)
	at script.s(browser.js:2)
	at script(browser.js:2)
	at script(browser.js:53038)
	at script.s(browser.js:2)
	at script(browser.js:2)
	at script(browser.js:53194)
	at script(browser.js:53193)
	at script.s(browser.js:2)
	at script(browser.js:2)
	at script(browser.js:45886)
	at script.s(browser.js:2)
	at script(browser.js:2)
	at script(browser.js:46294)
	at script(browser.js:46279)
	at script.s(browser.js:2)
	at script(browser.js:2)
	at script(browser.js:22728)
	at script.s(browser.js:2)
	at script(browser.js:2)
	at script(browser.js:21669)
	at script.s(browser.js:2)
	at script(browser.js:2)
	at script(browser.js:13974)
	at script.s(browser.js:2)
	at script(browser.js:2)
	at script(browser.js:5941)
	at script.s(browser.js:2)
	at script(browser.js:2)
	at script(browser.js:5787)
	at script(browser.js:5782)
	at script.s(browser.js:2)
	at script.e(browser.js:2)
	at script(browser.js:2)
	at script(browser.js:2)
	at script(browser.js:2)
	at org.mozilla.javascript.Interpreter.interpret(Interpreter.java:815)
	at org.mozilla.javascript.InterpretedFunction.call(InterpretedFunction.java:109)
	at org.mozilla.javascript.ContextFactory.doTopCall(ContextFactory.java:393)
	at org.mozilla.javascript.ScriptRuntime.doTopCall(ScriptRuntime.java:3282)
	at org.mozilla.javascript.InterpretedFunction.exec(InterpretedFunction.java:120)
	at org.mozilla.javascript.Context.evaluateReader(Context.java:1250)
	at org.mozilla.javascript.Context$evaluateReader$1.call(Unknown Source)
	at babel$_run_closure1.doCall(babel.groovy:13)
	at babel.run(babel.groovy:11)
```
If I wrap the code in `fromObject` in a try-catch block, it seems that the object in question is `null`. I'm still trying to collect more information.

### PR DESCRIPTION
**Pull request by gausie**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #2537 (sort of)
| Patch: Bug Fix?          | No
| Major: Breaking Change?  | No
| Minor: New Feature?      | 👍
| Tests Added + Pass?      | Not sure where they would go
| Documentation PR Link    | As far as I can see, we don't really need to add any
| Any Dependency Changes?  | No
| License                  | MIT

As suggested by @nicolo-ribaudo in https://github.com/mozilla/rhino/issues/661, I've updated `compat-table` to fully represent the latest version of Rhino and this PR should mean that we can now start targeting it with babel!

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/13448"></a>

### CODE REVIEW

**Comment by zloirock** [comment]
Could you run [`core-js-compat`](https://github.com/zloirock/core-js/tree/master/tests/compat) test case and add a PR to `core-js` too?

**Comment by gausie** [comment]
> Could you run [`core-js-compat`](https://github.com/zloirock/core-js/tree/master/tests/compat) test case and add a PR [to](https://github.com/zloirock/core-js/blob/master/packages/core-js-compat/src/data.js) `core-js` too?

Unfortunately Rhino can't really even run the test runner `core-js` uses (at least via my naive attempt). Would it be ok to merge this before getting that sorted or are they both needed concurrently?

**Comment by zloirock** [comment]
@gausie you could check the source of the test runner and adapt it to Rhino - it's just some lines. It's required for proper built-ins injection by `preset-env`.

**Comment by gausie** [comment]
@zloirock Sure, I'll adapt in my local copy for rhino, get the data and discard my edits.

**Comment by zloirock** [comment]
@gausie you could add a Rhino runner to the PR like to `compat-table` - it could simplify adding new data in the future -)

**Comment by gausie** [comment]
> @gausie you could add a Rhino runner to the PR like in `compat-table` - it could simplify adding new data in the future -)

I think I'll just fix the two issues in Rhino!

**Comment by gausie** [comment]
Here's that PR: https://github.com/zloirock/core-js/pull/942

**Comment by gausie** [comment]
@zloirock whats up with these two in-progress tests? Are they waiting for approval to start somehow?

**Comment by zloirock** [comment]
I think that it could be better to ask @nicolo-ribaudo since I'm not an expert in current CI.

**Review by zloirock — APPROVED** [review]

**Comment by nicolo-ribaudo** [comment]
We only run them manually when changing code that only affects Babel 8 or that might affect spec compliancy of our plugin.

**Review by nicolo-ribaudo — COMMENTED** [review]
Could you also add a test in `babel-preset-env/tests/fixtures` to make sure that the `rhino` target works?

**Comment by JLHwung** [comment]
We need to bump `core-js-compat` when https://github.com/zloirock/core-js/pull/942 is published.

**Comment by gausie** [comment]
@JLHwung that has now landed in core-js v3.15.0

**Comment by gausie** [comment]
Ok I've bumped `core-js-compat` but unsure if I should have bumped `core-js` too, so I've left it.

**Comment by nicolo-ribaudo** [comment]
@gausie I'm preparing a PR to bump all the `core-js` packages. I prefer to keep it separate from this one because this PR will be merged in the next minor, while bumping `core-js` is more urgent otherwise our e2e tests will start failing.

**Comment by gausie** [comment]
@nicolo-ribaudo ok, is bumping `core-js-compat` as I have done ok, or should I undo that last commit?

**Comment by nicolo-ribaudo** [comment]
We'll ask you to rebase this PR on `main` after https://github.com/babel/babel/pull/13491 is merged (or to merge `main` in this PR), so reverting the last commit is not necessary but might make it easier to rebase later.

**Comment by nicolo-ribaudo** [comment]
I just saw https://github.com/kangax/compat-table/pull/1731. Maybe for the same reason we should have Rhino patch versions in our data? We might need to update our build script to support them.

**Comment by gausie** [comment]
@nicolo-ribaudo Adding in that patch release is easy since every 1.7 would just need the same change, but I'm not sure how I'd go about updating the build scripts.

**Comment by nicolo-ribaudo** [comment]
Mmh, looking at our build script it looks like it should already work :thinking: Could you try updating the `COMPAT_TABLE` hash, and let's see if it works.

Btw, we just merged https://github.com/babel/babel/pull/13491 to `main`.

**Comment by gausie** [comment]
@nicolo-ribaudo Ok that's all done now

File: packages/babel-preset-env/test/fixtures/preset-options-babel-7/rhino-1_7/input.mjs
Comment (nicolo-ribaudo) [inline packages/babel-preset-env/test/fixtures/preset-options-babel-7/rhino-1_7/input.mjs:1]: Can you also add something that Rhino supports?

```suggestion
const a = () => 1;

0b01;
```

Also, this file should now be renamed to `1_7_3`

Comment (gausie) [reply packages/babel-preset-env/test/fixtures/preset-options-babel-7/rhino-1_7/input.mjs:1]: ah of course

Comment (gausie) [reply packages/babel-preset-env/test/fixtures/preset-options-babel-7/rhino-1_7/input.mjs:1]: well actually that's a great point i didn't even think about - rhino should support all of that so something is wrong here

Comment (nicolo-ribaudo) [reply packages/babel-preset-env/test/fixtures/preset-options-babel-7/rhino-1_7/input.mjs:1]: According to `compat-table`, it doesn't fully support arrow functions because it doesn't pass these tests:
- `lexical "super" binding in constructor`
- `lexical "super" binding in method`
- `lexical "new.target" bindin`

However, it doesn't pass them not because of arrow functions but because it doesn't support classes. Maybe we could exclude those three tests in https://github.com/babel/babel/blob/5145c98a735d8ea49fbaffd3d81033a6882bf570/packages/babel-compat-data/scripts/data/plugin-features.js#L35-L37 by explicitly listing all the other arrow tests? Similar to https://github.com/babel/babel/blob/5145c98a735d8ea49fbaffd3d81033a6882bf570/packages/babel-compat-data/scripts/data/plugin-features.js#L66-L73.

Those three removed tests can then be added to https://github.com/babel/babel/blob/5145c98a735d8ea49fbaffd3d81033a6882bf570/packages/babel-compat-data/scripts/data/plugin-features.js#L41-L43 and https://github.com/babel/babel/blob/5145c98a735d8ea49fbaffd3d81033a6882bf570/packages/babel-compat-data/scripts/data/plugin-features.js#L93-L95

Comment (gausie) [reply packages/babel-preset-env/test/fixtures/preset-options-babel-7/rhino-1_7/input.mjs]: Right that's all implemented now!

File: packages/babel-compat-data/scripts/data/plugin-features.js
Comment (nicolo-ribaudo) [inline packages/babel-compat-data/scripts/data/plugin-features.js:58]: We don't need the `new.target` one here, since it's compiled separately from classes.

File: packages/babel-compat-data/scripts/data/plugin-features.js
Comment (nicolo-ribaudo) [inline packages/babel-compat-data/scripts/data/plugin-features.js:115]: And we only need the `new.target` on here.

**Review by nicolo-ribaudo — COMMENTED** [review]
Sorry, I was not clear enough in my previous comment :sweat_smile:

File: packages/babel-compat-data/scripts/data/plugin-features.js
Comment (gausie) [reply packages/babel-compat-data/scripts/data/plugin-features.js:115]: Ah! Looking at it with that in mind it was sort of obvious haha

**Review by JLHwung — APPROVED** [review]
Does rhino support ES Modules? I searched its repo and it seems to me rhino only supports commonjs. If I am wrong, we should also update `packages/babel-compat-data/scripts/build-modules-support.js`, otherwise this PR lgtm.

**Comment by gausie** [comment]
> Does rhino support ES Modules? I searched its repo and it seems to me rhino only supports commonjs. If I am wrong, we should also update `packages/babel-compat-data/scripts/build-modules-support.js`, otherwise this PR lgtm.

You're correct, for now it only supports commonjs. Woop!

**Review by nicolo-ribaudo — APPROVED** [review]
Thanks! We'll merge this for the next minor release, somewhere in July.

**Comment by nicolo-ribaudo** [comment]
We'll release 7.15 this week. Thank you!

**Pull request merged**
